### PR TITLE
Add support for TinyUSB

### PR DESCRIPTION
**Pull request by Spegs21**
This creates other issues with missing identifiers due to differences in HID libraries and USBAPI.h between architectures.

### CODE REVIEW

**Comment by Spegs21** [comment]
RawHID.cpp

- `no instance of constructor "arduino::PluggableUSBModule::PluggableUSBModule" matches the argument list -- argument types are: (int, int, uint8_t [1])`
- `identifier "EP_TYPE_INTERRUPT_IN" is undefined`
- `identifier "TRANSFER_PGM" is undefined`
- `identifier "HID_REPORT_TYPE_FEATURE" is undefined`
- `identifier "HID_REPORT_TYPE_OUTPUT" is undefined`

RawHID.h

- `identifier "TRANSFER_RELEASE" is undefined`

**Comment by Spegs21** [comment]
I forgot the parenthesis and killed Uno/Mega sketches.

**Comment by Spegs21** [comment]
This now compiles for Arduino Zero (Native USB Port) when I comment out the EEPROM references. It might be kind of hacky but I didn't know how else to do it. You'll want to check it over.

**Comment by Spegs21** [comment]
I got this to build for a Adafruit Trinket M0 target but even more changes needed made, changes that conflict with the Arduino SAMD core.

Adafruit doesn't define Serial as a Uart interface. It's defined in their USBAPI.h instead. Also their PluggableUSBModule constructor wants a `uint32_t * epType` instead of `unsigned int *` like the Arduino SAMD core.

**Comment by Legion2** [comment]
I created an issue in the arduino core api repo https://github.com/arduino/ArduinoCore-API/issues/141

**Comment by Spegs21** [comment]
I also have a PR open for the missing HID defines in the SAMD core arduino/ArduinoCore-samd#602

**Comment by Spegs21** [comment]
I purchased a Trinket M0 and created the board definition for it. I was able to upload the sketch for Commander Core. Windows recognizes it as Commander Core when installing the driver but it doesn't show as Commander Core in the device manager. It shows as an 'HID compliant vender-defined device' but has the correct VID and PID.

I have not abstracted the EEPROM, so it's still disabled in the code. Do you think this has something to do with it? What do you think is the best approach to the abstraction?

**Comment by Legion2** [comment]
> What do you think is the best approach to the abstraction?

For `CorsairLightingFirmware` we need an interface (abstract class) which defines a getter and setter for DeviceId

For `FastLEDController` we need an interface for loading and storing the channels data

For both interfaces a EEPROM implementation should be provided which can be given in the constructor.
For the FastLEDController we can replace the `useEEPROM` flag with an empty implementation for the storage interface.

**Comment by Legion2** [comment]
@Spegs21 With "Commander Core" you mean LIGHTING NODE CORE? Because the Commander Core has a different PID and we are working on support in #211.

**Comment by Spegs21** [comment]
> @Spegs21 With "Commander Core" you mean LIGHTING NODE CORE? Because the Commander Core has a different PID and we are working on support in #211.

Yea, sorry. That was was a test of the Lighting Node Core example.

**Review by Legion2 — CHANGES_REQUESTED** [review]
please add the `samd` architecture to the libary.properties and if you can add an example samd board to the CI pipeline so we have a test compile to check that it works with the new architecture.

File: src/RawHID.h [hunk 1]
```diff
@@ -150,7 +165,7 @@ class RawHID_ : public PluggableUSBModule, public Stream {
 	bool setup(USBSetup& setup) override;
 	uint8_t getShortName(char* name) override;
 
-	EPTYPE_DESCRIPTOR_SIZE epType[ENDPOINT_COUNT];
+	EPTYPE_DESCRIPTOR_SIZE epType;
```
[the remark this answers is not in the document]
Comment (Spegs21) [reply src/RawHID.h:168]: Looks like I can. I will change it back.

File: src/RawHID.h [hunk 2]
```diff
@@ -51,9 +57,14 @@ THE SOFTWARE.
 
 #endif
 
-#if defined(ARDUINO_ARCH_AVR) && defined(USBCON)  // Arduino Core
+#if (defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_SAMD)) && defined(USBCON)  // Arduino Core
 
+#if defined(ARDUINO_ARCH_AVR)
 #define EPTYPE_DESCRIPTOR_SIZE uint8_t
+#else
+#define EPTYPE_DESCRIPTOR_SIZE unsigned int
```
[the remark this answers is not in the document]
Comment (Spegs21) [reply src/RawHID.h:65]: I tried but the compiler didn't like it.

File: src/RawHID.cpp [hunk 3]
```diff
@@ -154,7 +168,12 @@ bool CLP::RawHID_::setup(USBSetup& setup) {
 			else if (setup.wValueH == HID_REPORT_TYPE_OUTPUT) {
 				if (!dataAvailable && length <= dataLength) {
 					// Write data to fit to the end (not the beginning) of the array
+
+#if defined(ARDUINO_ARCH_AVR)
 					USB_RecvControl(data + dataLength - length, length);
+#else
+					USBDevice.recvControl(data + dataLength - length, length);
+#endif
```
[the remark this answers is not in the document]
Comment (Spegs21) [reply src/RawHID.cpp:176]: Sure, no problem.

File: src/RawHID.h [hunk 4]
```diff
@@ -25,11 +25,17 @@ THE SOFTWARE.
 
 #include <Arduino.h>
 
-#if defined(ARDUINO_ARCH_AVR)
+#if defined(ARDUINO_ARCH_SAMD)
+#define HID_REPORT_TYPE_INPUT 1
+#define HID_REPORT_TYPE_OUTPUT 2
+#define HID_REPORT_TYPE_FEATURE 3
+#endif
```
[the remark this answers is not in the document]
Comment (Spegs21) [reply src/RawHID.h:32]: I added the reference. Also changed to #ifndef statements too.

**Review by Legion2 — CHANGES_REQUESTED** [review]

**Comment by Spegs21** [comment]
I ended up forking the cmaglie/FlashStorage and added put and get templates which allow the original code to work with changes to which header is included. I tested it the library changes separately and they appear to work.

Unfortunately this didn't help. I'm still not seeing the device in iCUE. One things I've noticed in the iCUE logs is that vendor and model-name are not being reported even through these are set with the proper defines.

**Comment by Legion2** [comment]
For iCUE to detect the devices, you must define the USB IDs for the arduino boards. The boards are defined here https://github.com/Legion2/CorsairLightingProtocolBoards, we need to add also the samd boards

**Comment by Spegs21** [comment]
I'll add the boards definitions I created for the Trinket M0. I can create one for the Zero as well.

**Comment by Legion2** [comment]
It does not compile with the  CLP Adafruit board

**Comment by Spegs21** [comment]
There are a few reason for this:

- EPTYPE_DESCRIPTOR_SIZE differs between Arduino:Arduino SAMD and Adafruit:Arduino SAMD
I don't have a good solution for this yet.
- `extern Uart Serial` does not exist in the variant.h for the Trinket M0 and it cannot be added because they use it for `SERIAL_PORT_USBVIRTUAL`
I added `#ifndef USBCON` to CorsairLightingProtocolSerial.h/.cpp to work around this but I don't know if that's the right solution. Maybe we could pass a reference to the necessary serial to the CorsairLightingProtocolSerial constructor.

**Comment by Legion2** [comment]
Yes we can make it as constructor argument, but have to add it to the change log

**Comment by Spegs21** [comment]
> Yes we can make it as constructor argument, but have to add it to the changelog

Actual if we change `byte rawCommand[COMMAND_SIZE];` to `char rawCommand[COMMAND_SIZE];` in CorsairLightingProtocolSerial.h it seems to fix it.

**Comment by Legion2** [comment]
`byte` is the semantically the correct type here. I think we can exclude the file for the Trinket M0 board

**Comment by Legion2** [comment]
@Spegs21 have you tested this PR with the SAMD boards definitions we added in https://github.com/Legion2/CorsairLightingProtocolBoards/pull/7?

File: src/RawHID.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Legion2) [reply src/RawHID.h:71]: I added an extra `defined(ADAFRUIT_TRINKET_M0)` which should be a `defined(ARDUINO_SAMD_ADAFRUIT)`, but this is not set correctly in the CLP Boards config

**Comment by Legion2** [comment]
@XOIIO can you test these changes, you need to install the [CLP SAMD boards](https://github.com/Legion2/CorsairLightingProtocolBoards) and there is currenlty only a config for Arduino Zero and Adafruit Trinket M0, so you need to try one of them.

**Review by Legion2 — COMMENTED** [review]
@Spegs21 I added a storage interface for EEPROM which is optional so it works with SAMD

**Comment by Spegs21** [comment]
> @Spegs21 I added a storage interface for EEPROM which is optional so it works with SAMD

I'll check it  out tonight. Is there any cleanup I need to do on this PR? Sorry I kinda lost track of this.

**Comment by Legion2** [comment]
I already rebased the PR, cleaned the commits

**Comment by XOIIO** [comment]
@Legion2 unfortunately I do not have a zero or trinket on hand, only some pro minis, megas, and the seedstudio xiaos. I also didn't see either of those boards as an option after installing, but perhaps I misunderstood and those are the default ones before adding that board list.

https://imgur.com/gFk4eIr

**Comment by Spegs21** [comment]
@XOIIO You have to install `Corsair Lighting Protocol Boards - SAMD`.

**Comment by XOIIO** [comment]
Ah alright, thought I had clicked install but that didn't happen. I'll keep an eye open for when the definitions get updated and give it a test.

**Comment by Spegs21** [comment]
@Legion2 I tested the no EEPROM example but I'm running into the same issues I explained [here](https://github.com/Legion2/CorsairLightingProtocol/pull/215#issuecomment-783569324).
iCUE doesn't detect it. The manufacturer and device name don't show in the iCUE logs either, however the name is displayed by Windows when installing the driver.
I looked through the code but nothing jumped out at me as being incorrect.

**Comment by Legion2** [comment]
@Spegs21 can you upload a simple test sketch (blink) but using the CLP Board, so we can validate, that the CLP Boards definitions are working.

**Comment by Spegs21** [comment]
@Legion2 I rewrote this to use TinyUSB in a class that extends CLPResponse like your CLPHID class. I had some trouble at the start so I set up the NoEeprom example on vscode and made a SWD debugger out of a RPi Pico.

I've got it running now except I can't get it to respond. I've verified the response data is making it to my sendX method but after that I can't confirm what happens other than iCue doesn't see it and I can't get a response from it with and HID python script.

When I send the report I use report_id 0. It's seemed like you maybe send something different in RawHID. Is there a correct value to use here?

**Comment by Legion2** [comment]
I think I send it without an report_id. How did you setup the USB endpoints, how does your usb report look like?

**Comment by Spegs21** [comment]
I verified that my report descriptor I made with TinyUSB macros matches yours from RawHID, byte for byte. TinyUSB handles endpoints so I've just created an Adafruit_USBD_HID object and use that.

It's working better than the SAMD USB implementation did though. It's showing Vendor and Model Name in the iCue logs unlike it did before. It is also receiving the request for deviceID. 

The version of the TinyUSB library included in the Adafruit core doesn't allow for setting the serial number, newer versions do. I'm wondering is if that has something to do with it.

**Comment by Legion2** [comment]
I think the difference is, that I don't send a report, I just send the data here https://github.com/Legion2/CorsairLightingProtocol/blob/4593dca29f4844bca209e28032f863a7d31b2576/src/RawHID.h#L143 and do not use the SendReport function https://github.com/arduino/ArduinoCore-avr/blob/2cebe625833afcdb76cc941ccf90e5f5fefc27a9/libraries/HID/src/HID.cpp#L89-L96

**Comment by Legion2** [comment]
I have created an issue https://github.com/adafruit/Adafruit_TinyUSB_Arduino/issues/153

**Comment by Spegs21** [comment]
@Legion2 It looks like it already emulates this behavior by skipping the report_ID byte if it's 0.

**Comment by Legion2** [comment]
The endpoint does not matter, but the endpoint configuration is important

**Comment by Spegs21** [comment]
@Legion2 Finally got this working after installing Wireshark and seeing that the packets being sent from the device were considered malformed. The issue ended up being the HID report I was sending was 16 bytes but the report descriptor called for 64 bytes. I changed the descriptor to 16 and it worked right away.

I'll clean it up and update the PR.

**Comment by Spegs21** [comment]
@Legion2 I've made the TinyUSB implementation so that it no longer needs custom board definitions. VID, PID, manufacturer, and product are now set at runtime. Any board that is supported by TinyUSB should now work.

**Comment by Legion2** [comment]
@Spegs21 Is this MR ready for review? Then I can try to look into it tomorrow.

**Comment by Spegs21** [comment]
@Legion2 Yes, I think it's ready to go.

**Review by Legion2 — CHANGES_REQUESTED** [review]
This PR looks great, thanks for your work! Could you rebase it, so there are no merge commits.

**Comment by Spegs21** [comment]
@Legion2 I think I did this right, let me know if there's a problem with it.

**Comment by Spegs21** [comment]
@Legion2 I enabled workflows on my fork and have fixed any issues and squashed/rebased the commits.  It should pass the workflows here now.

File: src/RawHID.h [hunk 5]
```diff
@@ -23,10 +20,11 @@ THE SOFTWARE.
 */
 #pragma once
 
-#include <Arduino.h>
+#include "Arduino.h"
+#include "CorsairLightingProtocolConstants.h"
```
Comment (Legion2) [inline src/RawHID.h:24]: I don't like this reference here, RawHID should be decoupled from the Corsair specific stuff

Comment (Spegs21) [reply src/RawHID.h:24]: The issue was the serial number going into the constants. I use it in TinyUSBHID too. I didn't think we wanted to have it in two places. What do you want to do with it?

**Comment by Legion2** [comment]
@Spegs21 Thanks for your work, I will create a new release for it this week.

**Comment by Spegs21** [comment]
@Legion2 Awesome! Thanks for taking the time to review and incorporate these changes.

**Pull request merged**